crates/sui-types/src/multisig_legacy.rs

```diff
@@ -0,0 +1,364 @@
+        multisig_pk
+            .validate()
+            .map_err(|_| SuiError::InvalidSignature {
+                error: "Invalid multisig public key".to_string(),
+            })?;
+
+        if full_sigs.len() > multisig_pk.pk_map.len() || full_sigs.is_empty() {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid number of signatures".to_string(),
+            });
+        }
+        let mut bitmap = RoaringBitmap::new();
+        let mut sigs = Vec::with_capacity(full_sigs.len());
+        for s in full_sigs {
+            let pk = s.to_public_key()?;
+            let inserted = bitmap.insert(multisig_pk.get_index(&pk).ok_or(
+                SuiError::IncorrectSigner {
+                    error: format!("pk does not exist: {:?}", pk),
+                },
+            )?);
+            if !inserted {
+                return Err(SuiError::InvalidSignature {
+                    error: "Duplicate sigature".to_string(),
+                });
+            }
+            sigs.push(s.to_compressed()?);
+        }
+        Ok(MultiSigLegacy {
+            sigs,
+            bitmap,
+            multisig_pk,
+            bytes: OnceCell::new(),
+        })
+    }
+
+    pub fn validate(&self) -> Result<(), FastCryptoError> {
+        if self.sigs.len() > self.multisig_pk.pk_map.len() || self.sigs.is_empty() {
+            return Err(FastCryptoError::InvalidInput);
+        }
+        self.multisig_pk.validate()?;
+        Ok(())
+    }
+
+    pub fn get_pk(&self) -> &MultiSigPublicKeyLegacy {
+        &self.multisig_pk
+    }
+
+    pub fn get_sigs(&self) -> &[CompressedSignature] {
+        &self.sigs
+    }
+
+    pub fn get_bitmap(&self) -> &RoaringBitmap {
+        &self.bitmap
+    }
+}
+
+impl ToFromBytes for MultiSigLegacy {
+    fn from_bytes(bytes: &[u8]) -> Result<MultiSigLegacy, FastCryptoError> {
+        // The first byte matches the flag of MultiSig.
+        if bytes.first().ok_or(FastCryptoError::InvalidInput)? != &SignatureScheme::MultiSig.flag()
+        {
+            return Err(FastCryptoError::InvalidInput);
+        }
+        let multisig: MultiSigLegacy =
+            bcs::from_bytes(&bytes[1..]).map_err(|_| FastCryptoError::InvalidSignature)?;
+        multisig.validate()?;
+        Ok(multisig)
+    }
+}
+
+/// The struct that contains the public key used for authenticating a MultiSig.
+#[derive(Debug, Clone, PartialEq, Eq, JsonSchema, Serialize, Deserialize)]
+pub struct MultiSigPublicKeyLegacy {
+    /// A list of public key and its corresponding weight.
+    #[serde(serialize_with = "serialize_pk_map")]
+    #[serde(deserialize_with = "deserialize_pk_map")]
+    pk_map: Vec<(PublicKey, WeightUnit)>,
+    /// If the total weight of the public keys corresponding to verified signatures is larger than threshold, the MultiSig is verified.
+    threshold: ThresholdUnit,
+}
+
+fn serialize_pk_map<S>(pk_map: &[(PublicKey, WeightUnit)], serializer: S) -> Result<S::Ok, S::Error>
```

Comment:
pls add a comment describing why this serde is needed (legacy serde also for public keys)